### ISSUE TITLE
UnSHC a SHC binary that was not complied by relax security option

### ISSUE THREAD

**Issue by swatbelal** [yanncam/UnSHc#19]
Hello, 

long time ago (2-3 years ago) I had a dedicated that had a SHC encrypted shell script (script.sh.x file) And I do have that file til now but not that dedicated, it died long time ago, now I used SHC many many times and it always worked well, but when I complied that specific script I forgot to add the -r option to SHC command line "shc -f file.sh" ( -r is the relax security option to make a redistributable binary ), and now I really need to get that script to work again, I no longer have the source code of it, only the binary And whenever I run it on the other system it gives an error that says "abnormal behavior" and some other weird non English characters

Example:
root@[localhost]:~/shc # ./init.sh
./init.sh: ]��&͆��w�`jymM0}abnormal behavior!

When I tried using UnSHC it gave this result: (success)

root@[localhost]:~/shc # ./unshc.sh init.sh
 _   _       _____ _   _
| | | |     /  ___| | | |
| | | |_ __ \ `--.| |_| | ___
| | | | '_ \ `--. \  _  |/ __|
| |_| | | | /\__/ / | | | (__
 \___/|_| |_\____/\_| |_/\___|

--- UnSHc - The shc decrypter.
--- Version: 0.8
------------------------------
UnSHc is used to decrypt script encrypted with SHc
Original idea from Luiz Octavio Duarte (LOD)
Updated and modernized by Yann CAM
- SHc   : [http://www.datsi.fi.upm.es/~frosal/]
- UnSHc : [https://www.asafety.fr/unshc-the-shc-decrypter/]
------------------------------

[*] Input file name to decrypt [init.sh]
[+] ARC4 address call candidate : [0x400c74]
[*] Extracting each args address and size for the 14 arc4() calls with address [0x400c74]...
        [0] Working with var address at offset [0x602cf3] (0x2a bytes)
        [1] Working with var address at offset [0x602d1d] (0x1 bytes)
        [2] Working with var address at offset [0x602cdc] (0xa bytes)
        [3] Working with var address at offset [0x602d1e] (0x3 bytes)
        [4] Working with var address at offset [0x602d44] (0xf bytes)
        [5] Working with var address at offset [0x602d55] (0x1 bytes)
        [6] Working with var address at offset [0x602d27] (0x16 bytes)
        [7] Working with var address at offset [0x602d58] (0x16 bytes)
        [8] Working with var address at offset [0x602cc3] (0x13 bytes)
        [9] Working with var address at offset [0x602ce8] (0x1 bytes)
        [10] Working with var address at offset [0x602d21] (0x1 bytes)
        [11] Working with var address at offset [0x6023c4] (0x62d bytes)
        [12] Working with var address at offset [0x602caf] (0x13 bytes)
        [13] Working with var address at offset [0x602304] (0x13 bytes)
[*] Extracting password...
        [+] PWD address found : [0x602b92]
        [+] PWD size found : [0x100]
[*] Executing [/tmp/DdINn8] to decrypt [init.sh]
[*] Retrieving initial source code in [init.sh.sh]
[*] All done!

Meanwhile the output file only had this:

root@[localhost]:~/shc # cat init.sh.sh
젢3����3E���Q�4��Mj�s�=ϣ^�Y␋�2�YIS����M�≠��     �"�─J���␉┤π1���≠7��������
                                                                         %N�Q⬥�(<�������ل��񚬤␋%ݪ���H��Б▲���F���█�����Ƚ^֏├�����@�� ����(≠
                                     ��W�࢚����O�⎽�┌�%�&��>��ݢ��Xۙ����␍4���]����������)ۭ���≥≠^␊ ��⎼⎺⎺├@[≥G▲S├▒├┤⎽]:•/⎽#␌ #

Any ideas? Thanks in advance.

**Comment by swatbelal** [comment]
Anyone?

**Comment by yanncam** [comment]
Hello,

Sorry for the late response.
I haven't any good solution for this case where the `-r` option was missing initialy.
UnSHc doesn't seem to be able to recover this kind of encrypted file not-relaxed.

Unfortunately I do not have time right now to dig into this problem.

But I encourage you to try to reverse manually your binary, to understand precisly "what do SHc without the `-r` parameter ?".

Recently I've written an article in a french-cybersecurity oriented magazine about UnSHc, with all the detailled process to do it manually. You can try this way : 

* Part 1 : https://www.miscmag.com/unshc-dechiffrer-des-scripts-sh-compiles-et-chiffres-par-shc-partie-12/
* Part 2 : https://www.miscmag.com/unshc-dechiffrer-des-scripts-sh-compiles-et-chiffres-par-shc-partie-22/

Sincerely,

**Comment by swatbelal** [comment]
I'll see, thanks

**Comment by intika** [comment]
SHC relax feature (variable `rlax` on the sources) does use as an encryption/decryption the attributes of the file (`/usr/bin/sh` or `/usr/bin/bash` or else is your system use a different shell), by attributes i mean file size, date etc.

```
	memset(control, 0, sizeof(control));
	control->st_ino = statf->st_ino;
	control->st_dev = statf->st_dev;
	control->st_rdev = statf->st_rdev;
	control->st_uid = statf->st_uid;
	control->st_gid = statf->st_gid;
	control->st_size = statf->st_size;
	control->st_mtime = statf->st_mtime;
	control->st_ctime = statf->st_ctime;
```
Control variable is then used to generate the key... so if you enable the the `-r` option that key won't be used otherwise your shell executable need to be the exact same while you built your sh script (even `ctime`, not just the file version)

In short in order for you to recover your file you need to restore `/usr/bin/bash` or `/usr/bin/sh` with the exact same attributes as the time you built your sh script on.

Additional [note](https://unix.stackexchange.com/a/252017/120919).
